pkg/controllermanager/controller/project/resourcequota/add.go

```diff
@@ -0,0 +1,103 @@
+	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "project-resourcequota"
+
+var (
+	createOnlyPredicate = predicate.Funcs{
+		CreateFunc:  func(_ event.CreateEvent) bool { return true },
+		DeleteFunc:  func(_ event.DeleteEvent) bool { return false },
+		UpdateFunc:  func(_ event.UpdateEvent) bool { return false },
+		GenericFunc: func(_ event.GenericEvent) bool { return false },
+	}
+)
+
+// AddToManager adds a controller with the given Options to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.ConcurrentSyncs, 0),
+			ReconciliationTimeout:   controllerutils.DefaultReconciliationTimeout,
+		}).
+		For(&corev1.ResourceQuota{}, builder.WithPredicates(r.ObjectInProjectNamespace())).
+		Watches(
+			&v1beta1.Shoot{},
+			handler.EnqueueRequestsFromMapFunc(r.MapShootToResourceQuotasInProject(mgr.GetLogger().WithValues("controller", ControllerName))),
+			builder.WithPredicates(createOnlyPredicate),
+		).
+		Complete(r)
+}
+
+// ObjectInProjectNamespace returns a predicate that filters objects that are in Project namespaces.
+func (r *Reconciler) ObjectInProjectNamespace() predicate.Predicate {
+	objectInNamespaceFunc := func(namespace string) bool {
+		ctx := context.Background()
+		project, err := gardenerutils.ProjectForNamespaceFromReader(ctx, r.Client, namespace)
+		return err == nil && project != nil
+	}
+
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return objectInNamespaceFunc(e.Object.GetNamespace())
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return objectInNamespaceFunc(e.Object.GetNamespace())
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return objectInNamespaceFunc(e.ObjectNew.GetNamespace())
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return objectInNamespaceFunc(e.Object.GetNamespace())
+		},
+	}
+}
+
+// MapShootToResourceQuotasInProject maps Shoots to ResourceQuotas in the corresponding Project namespace.
+func (r *Reconciler) MapShootToResourceQuotasInProject(log logr.Logger) handler.MapFunc {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		resourceQuotaList := &corev1.ResourceQuotaList{}
+		if err := r.Client.List(ctx, resourceQuotaList, client.InNamespace(obj.GetNamespace())); err != nil {
+			log.Error(err, "Unable to list resource quotas")
+		}
+
+		requests := make([]reconcile.Request, 0, len(resourceQuotaList.Items))
+		for _, rq := range resourceQuotaList.Items {
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+				Name:      rq.Name,
+				Namespace: rq.Namespace,
+			}})
+		}
+		return requests
+	}
+}
```

Comment:
You can simplify this, see e.g. https://github.com/gardener/gardener/blob/c4d228456f95d54608320fd102737fc3fa8e480a/pkg/controllermanager/controller/controllerregistration/seed/add.go#L248-L259